azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkits/appservice/service/impl/WebAppDeploymentSlot.java

```diff
@@ -34,81 +39,163 @@ public WebAppDeploymentSlot(WebAppDeploymentSlotEntity deploymentSlot, AzureAppS
     }
 
     @Override
     public void deploy(File file) {
-
+        deploy(AppServiceUtils.getDeployTypeByFileExtension(file), file);
     }
 
     @Override
     public void deploy(DeployType deployType, File file) {
-
+        getDeploymentSlotClient().deploy(com.azure.resourcemanager.appservice.models.DeployType.fromString(deployType.getValue()), file);
     }
 
     @Override
     public boolean exists() {
-        return false;
+        return getDeploymentSlotClient(true) != null;
     }
 
     @Override
     public String hostName() {
-        return null;
+        return getDeploymentSlotClient().defaultHostname();
     }
 
     @Override
     public String state() {
-        return null;
+        return getDeploymentSlotClient().state();
     }
 
     @Override
     public Runtime getRuntime() {
-        return null;
+        return AppServiceUtils.getRuntimeFromWebApp(getDeploymentSlotClient());
     }
 
     @Override
     public PublishingProfile getPublishingProfile() {
-        return null;
+        return AppServiceUtils.getPublishingProfile(getDeploymentSlotClient().getPublishingProfile());
+    }
+
+    @Override
+    public WebAppDeploymentSlotEntity entity() {
+        return slotEntity;
+    }
+
+    private com.azure.resourcemanager.appservice.models.DeploymentSlot getDeploymentSlotClient() {
+        return getDeploymentSlotClient(false);
+    }
+
+    private synchronized com.azure.resourcemanager.appservice.models.DeploymentSlot getDeploymentSlotClient(boolean force) {
+        if (deploymentSlotClient == null || force) {
+            try {
+                final WebApp webAppService = StringUtils.isNotEmpty(slotEntity.getId()) ?
+                        azureClient.webApps().getById(slotEntity.getId().substring(0, slotEntity.getId().indexOf("/slots"))) :
+                        azureClient.webApps().getByResourceGroup(slotEntity.getResourceGroup(), slotEntity.getWebappName());
+                deploymentSlotClient = StringUtils.isNotEmpty(slotEntity.getId()) ? webAppService.deploymentSlots().getById(slotEntity.getId()) :
+                        webAppService.deploymentSlots().getByName(slotEntity.getName());
+                slotEntity = AppServiceUtils.getWebAppDeploymentSlotEntity(deploymentSlotClient);
+            } catch (Exception e) {
+                // SDK will throw exception when resource not founded
+                return null;
+            }
+        }
+        return deploymentSlotClient;
     }
 
     @Override
     public String id() {
-        return null;
+        return getDeploymentSlotClient().id();
     }
 
     @Override
     public String name() {
-        return null;
-    }
-}
\ No newline at end of file
+        return getDeploymentSlotClient().name();
+    }
+
+    @Getter
+    public class WebAppDeploymentSlotCreator implements IWebAppDeploymentSlotCreator {
+        public static final String CONFIGURATION_SOURCE_NEW = "new";
```

Comment:
priavte

Reply:
They are designed to be public for tooling.